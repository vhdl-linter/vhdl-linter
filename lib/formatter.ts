class Formatter {
    private pos: number = 0;
    constructor(public text: string) {

    }
    gotoChar(pos: number) {
        if (pos === -1) {
            this.pos = this.text.length - 1;
        } else {
            this.pos = pos;
        }
    }
    vhdlBeautifyRegion() {
        this.vhdlFixupWhitespaceRegion();
        // this.vhdlFixStatementRegion();
        // this.vhdlIndentRegion();
        // this.vhdlAlignRegion();
        // this.vhdlFixCaseRegion();
        // this.vhdlRemoveTrailingSpacesRegion();
        // Fix Tabs
        return this.text;
    }
    vhdlFixStatementRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlIndentRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlAlignRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlFixCaseRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlRemoveTrailingSpacesRegion() {
        throw new Error("Method not implemented.");
    }
    match: RegExpExecArray;
    matchStart: number;
    matchEnd: number;
    reSearchForward(re: RegExp | string) {
        if (typeof re === 'string') {
            re = new RegExp(re);
        }
        const match = re.exec(this.text.substring(this.pos));
        if (match === null) {
            return false;
        }
        this.matchStart = this.pos + match.index;
        this.matchEnd = this.matchStart + match[0].length;
        return true;
    }
    replaceMatch(replacement: string, subgroup?: number) {
        if (subgroup === undefined) {
            this.text = this.text.substring(0, this.matchStart) + replacement + this.text.substring(this.matchEnd);
        } else {
            let newReplacement = '';
            for (let i = 0; i < this.match.length; i++) {
                if (i === subgroup) {
                    newReplacement += replacement;
                } else {
                    newReplacement += this.match[i];
                }
            }
            this.text = this.text.substring(0, this.matchStart) + newReplacement + this.text.substring(this.matchEnd);
        }
        // TODO: Correct Cursor?
    }
    vhdlFixupWhitespaceRegion() {
        //         (goto - char end)
        //         (setq end(point - marker))


        //         ;; have no space before and one space after`,' and ';'
        //     (goto-char beg)
        this.gotoChar(0);
        //     (while (re-search-forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\(\\s-*\\([,;]\\)\\)" end t)
        //       (if (match-string 1)
        // 	  (goto-char (match-end 1))
        // 	(replace-match "\\3 " nil nil nil 2)))
        while (this.reSearchForward("\\(--.*\\n\\|\"[^\"\\n]*[\"\\n]\\|'.'\\|\\\\[^\\n]*[\\n]\\)\\|\\(\\s-*\\([,;]\\)\\)")) {
            this.gotoChar(this.matchEnd);
            this.replaceMatch(this.match[3], 2);
        }
        //     ;; have no space after `('
        //             (goto - char beg)
        this.gotoChar(0);
        //             (while (re - search - forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\((\\)\\s-+" end t)
        //         (if (match - string 1)
        //         (goto - char(match - end 1))
        //             (replace - match "\\2")))
        // while (this.reSearchForward("\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\((\\)\\s-+")) {
        //     this.gotoChar(this.matchEnd);
        //     this.replaceMatch(this.match[3], 2);
        // }



        //         ;; have no space before `)'
        //     (goto-char beg)
        //     (while (re-search-forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\|^\\s-+\\)\\|\\s-+\\()\\)" end t)
        //       (if (match-string 1)
        // 	  (goto-char (match-end 1))
        // 	(replace-match "\\2")))
        //     ;; surround operator symbols by one space
        //     (goto-char beg)
        //     (while (re-search-forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\(\\([^/:<>=\n]\\)\\(:\\|\\??=\\|\\??<<\\|\\??>>\\|\\??<\\|\\??>\\|:=\\|\\??<=\\|\\??>=\\|=>\\|\\??/=\\|\\?\\?\\)\\([^=>\n]\\|$\\)\\)" end t)
        //       (if (or (match-string 1)
        // 	      (<= (match-beginning 0)  ; not if at boi
        // 		 (save-excursion (back-to-indentation) (point))))
        // 	  (goto-char (match-end 0))
        // 	(replace-match "\\3 \\4 \\5")
        // 	(goto-char (match-end 2))))
        //     ;; eliminate multiple spaces and spaces at end of line
        //     (goto-char beg)
        //     (while (or (and (looking-at "--.*\n") (re-search-forward "--.*\n" end t))
        // 	       (and (looking-at "--.*") (re-search-forward "--.*" end t))
        // 	       (and (looking-at "\"") (re-search-forward "\"[^\"\n]*[\"\n]" end t))
        // 	       (and (looking-at "\\s-+$") (re-search-forward "\\s-+$" end t)
        // 		    (progn (replace-match "" nil nil) t))
        // 	       (and (looking-at "\\s-+;") (re-search-forward "\\s-+;" end t)
        // 		    (progn (replace-match ";" nil nil) t))
        // 	       (and (looking-at "^\\s-+") (re-search-forward "^\\s-+" end t))
        // 	       (and (looking-at "\\s-+--") (re-search-forward "\\s-+" end t)
        // 		    (progn (replace-match "  " nil nil) t))
        // 	       (and (looking-at "\\s-+") (re-search-forward "\\s-+" end t)
        // 		    (progn (replace-match " " nil nil) t))
        // 	       (and (looking-at "-") (re-search-forward "-" end t))
        // 	       (re-search-forward "[^ \t\"-]+" end t))))
        //   (unless no-message (message "Fixing up whitespace...done")))
    }
}

const test = `asdasd,asdasd
asdasdasd ; 
asdasd`;
const formatter = new Formatter(test);
formatter.vhdlBeautifyRegion();
console.log(formatter.text);