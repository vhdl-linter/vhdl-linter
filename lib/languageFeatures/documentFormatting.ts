import { exec, spawn } from 'child_process';
import { promises } from 'fs';
import { tmpdir } from 'os';
import { sep } from 'path';
import { platform } from 'process';
import { fileURLToPath } from 'url';
import { promisify } from 'util';
import { CancellationToken, DocumentFormattingParams, LSPErrorCodes, Range, ResponseError, TextEdit, WorkDoneProgressReporter } from 'vscode-languageserver';
import { attachWorkDone } from 'vscode-languageserver/lib/common/progress';
import { connection, documents } from '../language-server';
import { getRootDirectory, joinURL } from '../project-parser';
// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-argument
const nullProgressReporter = attachWorkDone(undefined as any, /* params */ undefined);
async function _getProgressReporter(reporter: WorkDoneProgressReporter, title: string) {
  // This is a bit ugly, but we need to determine whether the provided reporter
  // is an actual client-side progress reporter or a dummy (null) progress reporter
  // created by the LSP library. If it's the latter, we'll create a server-initiated
  // progress reporter.
  if (reporter.constructor !== nullProgressReporter.constructor) {
    return reporter;
  }

  const serverInitiatedReporter = await connection.window.createWorkDoneProgress();
  serverInitiatedReporter.begin(
    title
  );

  return serverInitiatedReporter;
}
const execP = promisify(exec);
export async function handleDocumentFormatting(params: DocumentFormattingParams, token: CancellationToken, workDoneProgress: WorkDoneProgressReporter): Promise<TextEdit[] | null | ResponseError> {
  const document = documents.get(params.textDocument.uri);
  if (typeof document === 'undefined') {
    return null;
  }
  const text = document.getText();
  const path = await promises.mkdtemp(tmpdir() + sep);
  const tmpFile = path + sep + 'beautify';
  await promises.writeFile(tmpFile, text);
  const rootDir = getRootDirectory();
  const emacsScripts = joinURL(rootDir, 'emacs', 'emacs-vhdl-formatting-script.lisp');
  const emacsLoadPath = joinURL(rootDir, 'emacs');
  const numSpaces = typeof params.options.tabSize === 'number' ? params.options.tabSize : 2;
  let foundEmacs = false;
  if (platform === 'win32') {
    try {
      await execP(`emacs --batch`);
      foundEmacs = true;
    } catch (e) {
    }
  } else {
    try {
      await execP(`command -v emacs`);
      foundEmacs = true;
    } catch (e) {
    }
  }
  if (!foundEmacs) {
    connection.window.showErrorMessage('vhdl-linter is using emacs for formatting. Install emacs for formatting to work.');
    return null;

  }
  const controller = new AbortController();
  const { signal } = controller;
  token.onCancellationRequested(() => {
    controller.abort();
  });
  const progress = await _getProgressReporter(workDoneProgress, 'Emacs Formatter running');

  if (platform === 'win32') {
    progress.report(0, "Emacs Formatter running");

    const cmd = `emacs --batch --eval "(setq-default vhdl-basic-offset ${numSpaces})" ` +
      // The vhdl-mode from the folder does not work with win for some reason
      // `--eval "(setq load-path (cons (expand-file-name \\"${fileURLToPath(emacsLoadPath)}\\") load-path))" ` +
      ` -l ${fileURLToPath(emacsScripts)} -f vhdl-batch-indent-region ${tmpFile}`
    try {
      await execP(cmd);
    } catch (e) {
      progress.done();

      if (e.name === "AbortError") {
        return new ResponseError(LSPErrorCodes.RequestCancelled, 'canceled');
      }
      throw e;
    }
    progress.done();

  } else {
    try {
      await new Promise<void>((resolve, reject) => {
        const args = ['-c', `emacs --batch --eval "(setq-default vhdl-basic-offset ${numSpaces})" ` +
          `--eval "(setq load-path (cons (expand-file-name \\"${fileURLToPath(emacsLoadPath)}\\") load-path))" ` +
          ` -l ${fileURLToPath(emacsScripts)} -f vhdl-batch-indent-region ${tmpFile}`];
        const emacs = spawn('sh', args, { signal });

        emacs.stderr.on('data', (data: Buffer) => {
          const match = data.toString().match(/(\d+)%/) as [string, string] | null;
          if (match) {
            progress.report(parseInt(match[1]), data.toString());
          }
        });
        emacs.on('error', err => reject(err));
        emacs.on('close', code => {
          if (code === 0) {
            resolve();
          }
          reject(code);
        });
      });

      progress.done();
    } catch (e) {
      progress.done();

      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      if (e.name === "AbortError") {
        return new ResponseError(LSPErrorCodes.RequestCancelled, 'canceled');
      }
      throw e;
    }
  }
  // Emacs VHDL mode hard tabs seem to be broken.
  // Replace spaces in beginning with tabs
  let newText = await promises.readFile(tmpFile, { encoding: 'utf8' });
  if (params.options.insertSpaces === false) {
    const re = /^( +)/gm;
    console.log(re);
    newText = newText.replaceAll(re,
      match => '\t'.repeat(Math.ceil(match.length / numSpaces))
    );
  }
  return [{
    range: Range.create(document.positionAt(0), document.positionAt(text.length)),
    newText
  }];
}