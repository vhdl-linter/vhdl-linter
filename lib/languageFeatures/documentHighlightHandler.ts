import { DocumentHighlight, DocumentHighlightKind, TextDocumentPositionParams } from 'vscode-languageserver';
import { OLexerToken } from '../lexer';
import { IHasLexerToken, IHasNameToken, implementsIHasLexerToken, implementsIHasNameToken } from '../parser/interfaces';
import { ObjectBase, OName } from '../objects/objectsIndex';
import { VhdlLinter } from '../vhdlLinter';
import { findObjectFromPosition } from './findObjects';

export function documentHighlightHandler(linter: VhdlLinter, params: TextDocumentPositionParams): DocumentHighlight[] {
  const startI = linter.getIFromPosition(params.position);

  const candidates = findObjectFromPosition(linter, params.position)
    .filter(candidate => {
      // Only use Objects that have a lexer Token (either directly or as a reference Token)
      // Also check if the cursor is actually in that token itself, or just in that objects range.
      if (implementsIHasLexerToken(candidate)) {
        if (candidate.lexerToken.range.start.i <= startI && startI <= candidate.lexerToken.range.end.i) {
          return true;
        }
      } else if (implementsIHasNameToken(candidate)) {
        if (candidate.nameToken.range.start.i <= startI && startI <= candidate.nameToken.range.end.i) {
          return true;
        }
      }
      return false;

    }) as (ObjectBase & (IHasLexerToken | IHasNameToken))[];

  if (candidates.length === 0) {
    return [];
  }
  const candidate = candidates[0];
  const highlights: DocumentHighlight[] = [];

  const searchForToken = (token: OLexerToken) => {
    const name = token.getLText();
    for (const object of linter.file.objectList) {
      if (implementsIHasLexerToken(object) && object.lexerToken.getLText() === name) {
        highlights.push({
          range: object.lexerToken.range,
          kind: DocumentHighlightKind.Read
        });
      } else if (implementsIHasNameToken(object) && object.nameToken.getLText() === name) {
        highlights.push({
          range: object.nameToken.range,
          kind: object instanceof OName && object.write ? DocumentHighlightKind.Write : DocumentHighlightKind.Read
        });
      }
    }
  };
  if (candidate) {
    if (implementsIHasLexerToken(candidate)) {
      searchForToken(candidate.lexerToken);
    }
    if (implementsIHasNameToken(candidate)) {
      searchForToken(candidate.nameToken);
    }
  }
  return highlights;

}