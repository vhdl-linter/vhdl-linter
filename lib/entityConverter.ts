import { Position } from 'vscode-languageserver';
import { findObjectFromPosition } from './languageFeatures/findObjects';
import { OArchitecture, OEntity, OGeneric, OPort } from './parser/objects';
import { ISettings } from './settingsGenerated';
import { VhdlLinter } from './vhdlLinter';


function longestInArray(array: OPort[] | OGeneric[]) {
    let longest = 0;
    for (const object of array) {
        if (object.lexerToken.text.length > longest) {
            longest = object.lexerToken.text.length;
        }
    }
    return longest;
}
export function instanceTemplate(entity: OEntity, settings: ISettings) {
    let text = `${settings.style.instantiationLabelPrefix}${entity.lexerToken.text}${settings.style.instantiationLabelSuffix} : entity work.${entity.lexerToken.text}`;
    const indentString = '  ';
    if (entity.generics.length > 0) {
        text += `\ngeneric map (\n`;
        const longest = longestInArray(entity.generics);
        for (const generic of entity.generics) {
            const name = generic.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}${name} => ${settings.style.genericPrefix}${generic.lexerToken.text}${settings.style.genericSuffix},\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    if (entity.ports.length > 0) {
        text += `\nport map (\n`;
        const longest = longestInArray(entity.ports);
        for (const port of entity.ports) {
            const name = port.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}${name} => ${settings.style.signalPrefix}${port.lexerToken.text}${settings.style.signalSuffix},\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    text += `;\n`;
    return text;
}
export function componentTemplate(entity: OEntity) {
    let text = `component ${entity.lexerToken.text} is`;
    const indentString = '  ';
    if (entity.generics.length > 0) {
        text += `\ngeneric (\n`;
        for (const generic of entity.generics) {
            text += `${indentString}${generic.range.getText()};\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n);`;
    }

    if (entity.ports.length > 0) {
        text += `\nport (\n`;
        for (const port of entity.ports) {
            text += `${indentString}${port.range.getText()};\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n);`;
    }

    text += `\nend component;\n`;
    return text;
}

export function sysVerilogTemplate(entity: OEntity, settings: ISettings) {
    let text = entity.lexerToken.text;
    const indentString = '  ';
    if (entity.generics.length > 0) {
        text += ` #(\n`;
        const longest = longestInArray(entity.generics);
        for (const generic of entity.generics) {
            const name = generic.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}.${name}(${settings.style.genericPrefix}${generic.lexerToken.text}${settings.style.genericSuffix}),\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    if (entity.ports.length > 0) {
        text += ` ${settings.style.instantiationLabelPrefix}${entity.lexerToken.text}${settings.style.instantiationLabelSuffix} (\n`;
        const longest = longestInArray(entity.ports);
        for (const port of entity.ports) {
            const name = port.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}.${name}(${settings.style.signalPrefix}${port.lexerToken.text}${settings.style.signalSuffix}),\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    text += `;\n`;
    return text;
}
export function signalsTemplate(entity: OEntity, settings: ISettings) {
    let text = '';
    if (entity.ports.length > 0) {
        const longest = longestInArray(entity.ports);
        for (const port of entity.ports) {
            const name = port.lexerToken.text.padEnd(longest, ' ');
            let typeText: string | undefined;
            if (port.subtypeIndication.typeNames.length > 0) {
                typeText = port.subtypeIndication.typeNames[0]!.range.copyWithNewEnd(port.range.end).getText();
            }
            text += `signal ${settings.style.signalPrefix}${name}${settings.style.signalSuffix} : ${typeText ?? 'unknown_type'};\n`;
        }
    }
    return text;
}
export type converterTypes = 'instance' | 'signals' | 'sysverilog' | 'component';
export function entityConverter(vhdlLinter: VhdlLinter, type: converterTypes, settings: ISettings, position?: Position) {
    let entity;
    if (position === undefined) {
        entity = vhdlLinter.file.entities[0];
    } else {
        const rootElement = findObjectFromPosition(vhdlLinter, position)[0]?.getRootElement();
        if (rootElement instanceof OEntity) {
            entity = rootElement;
        } else if (rootElement instanceof(OArchitecture) && rootElement.correspondingEntity) {
            entity = rootElement.correspondingEntity;
        } else {
            // Fallback to first entity
            entity = vhdlLinter.file.entities[0];
        }
    }
    if (!entity) {
        return;
    }
    if (type === 'instance') {
        return instanceTemplate(entity, settings);
    } else if (type === 'signals') {
        return signalsTemplate(entity, settings);
    } else if (type === 'sysverilog') {
        return sysVerilogTemplate(entity, settings);
    } else if (type === 'component') {
        return componentTemplate(entity);
    }
}