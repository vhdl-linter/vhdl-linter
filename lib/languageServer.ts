import { existsSync } from 'fs';
import { pathToFileURL } from 'url';
import { TextDocument } from 'vscode-languageserver-textdocument';
import {
  CancellationToken,
  CodeAction, createConnection, DidChangeConfigurationNotification, InitializeParams, LSPErrorCodes, Position, ProposedFeatures, ResponseError, TextDocuments, TextDocumentSyncKind
} from 'vscode-languageserver/node';
import { getCompletions } from './languageFeatures/completion';
import { handleDocumentFormatting } from './languageFeatures/documentFormatting';
import { documentHighlightHandler } from './languageFeatures/documentHighlightHandler';
import { DocumentSymbols } from './languageFeatures/documentSymbol';
import { findDefinitionLinks } from './languageFeatures/findDefinition';
import { findReferencesHandler } from './languageFeatures/findReferencesHandler';
import { foldingHandler } from './languageFeatures/folding';
import { prepareRenameHandler, renameHandler } from './languageFeatures/rename';
import { semanticToken, semanticTokensLegend } from './languageFeatures/semanticToken';
import { signatureHelp } from './languageFeatures/signatureHelp';
import { workspaceSymbol } from './languageFeatures/workspaceSymbol';
import { LinterManager } from './linterManager';
import { normalizeUri } from './normalizeUri';
import { ProjectParser } from './projectParser';
import { defaultSettings, ISettings, normalizeSettings } from './settings';
import { entityConverter, converterTypes } from './entityConverter';

// Create a connection for the server. The connection auto detected protocol
// Also include all preview / proposed LSP features.
export const connection = createConnection(ProposedFeatures.all);


// Create a simple text document manager. The text document manager
// supports full document sync only
export const documents = new TextDocuments<TextDocument>(TextDocument);

let hasWorkspaceFolderCapability = false;
// let hasDiagnosticRelatedInformationCapability: boolean = false;
let projectParser: ProjectParser;
let rootUri: string | undefined;




let globalSettings: ISettings = defaultSettings;
let hasConfigurationCapability = false;
// Cache the settings of all open documents
const documentSettings = new Map<string, ISettings>();
connection.onDidChangeConfiguration((change: { settings: { VhdlLinter?: ISettings } }) => {

  if (hasConfigurationCapability) {
    // Reset all cached document settings
    documentSettings.clear();
  } else {
    globalSettings = (
      (change.settings.VhdlLinter ?? defaultSettings)
    );
  }

  // Revalidate all open text documents
  for (const document of documents.all()) {
    void validateTextDocument(document);
  }
});
export async function getDocumentSettings(resource?: URL): Promise<ISettings> {
  if (!hasConfigurationCapability) {
    return normalizeSettings(globalSettings);
  }
  let result = documentSettings.get(resource?.toString() ?? '');
  if (!result) {
    result = normalizeSettings(await connection.workspace.getConfiguration({
      scopeUri: resource?.toString(),
      section: 'VhdlLinter'
    }) as ISettings);
  }
  documentSettings.set(resource?.toString() ?? '', result);
  return result;
}
connection.onInitialize((params: InitializeParams) => {
  const capabilities = params.capabilities;
  hasWorkspaceFolderCapability =
    capabilities.workspace?.workspaceFolders ?? false;
  if (params.rootUri !== null) {
    rootUri = params.rootUri;
  }
  hasConfigurationCapability = capabilities.workspace?.configuration ?? false;
  return {
    capabilities: {
      textDocumentSync: TextDocumentSyncKind.Full,
      codeActionProvider: true,
      // Tell the client that the server supports code completion
      completionProvider: {
        resolveProvider: false
      },
      semanticTokensProvider: {
        legend: semanticTokensLegend,
        range: false,
        full: {
          delta: false
        }
      },

      documentSymbolProvider: true,
      definitionProvider: true,
      hoverProvider: true,
      documentFormattingProvider: {
        workDoneProgress: true
      },
      referencesProvider: true,
      foldingRangeProvider: true,
      documentHighlightProvider: true,
      executeCommandProvider: { commands: ['vhdl-linter:lsp-command'] },
      renameProvider: {
        prepareProvider: true
      },
      workspaceSymbolProvider: true,
      signatureHelpProvider: {
        triggerCharacters: [
          '(', ',', '>'
        ]
      }
    }
  };
});
export const initialization = new Promise<void>(resolve => {
  connection.onInitialized(() => {

    const handler = async () => {
      const progress = await connection.window.createWorkDoneProgress();
      progress.begin(
        'VHDL-linter initializing...',
        0
      );
      if (hasConfigurationCapability) {
        // Register for all configuration changes.
        await connection.client.register(DidChangeConfigurationNotification.type, undefined);
      }
      const configuration = (await connection.workspace.getConfiguration({
        section: 'VhdlLinter'
      })) as ISettings;

      if (hasWorkspaceFolderCapability) {
        const parseWorkspaces = async () => {
          const workspaceFolders = await connection.workspace.getWorkspaceFolders();
          const folders = (workspaceFolders ?? []).map(workspaceFolder => new URL(workspaceFolder.uri));
          folders.push(...configuration.paths.additional.map(path => pathToFileURL(path))
            .filter(url => existsSync(url)));

          projectParser = await ProjectParser.create(folders, configuration.paths.ignoreRegex, getDocumentSettings, false, progress);
        };
        await parseWorkspaces();
        connection.workspace.onDidChangeWorkspaceFolders(event => {
          projectParser.addFolders(event.added.map(folder => new URL(folder.uri)));
          connection.console.log('Workspace folder change event received.');
        });
      } else {
        const folders = [];
        if (rootUri !== undefined) {
          folders.push(new URL(rootUri));
        }
        projectParser = await ProjectParser.create(folders, configuration.paths.ignoreRegex, getDocumentSettings, false, progress);
      }
      for (const textDocument of documents.all()) {
        await validateTextDocument(textDocument);
      }
      projectParser.events.on('change', (_, uri: string) => {
        // A file in project parser got changed
        // Revalidate all *other* files. (The file itself gets directly handled.)
        for (const document of documents.all()) {
          if (normalizeUri(document.uri) !== uri) {
            void validateTextDocument(document, true);
          }
        }
      });
      documents.onDidChangeContent(change => {
        void validateTextDocument(change.document);
      });
      progress.done();
      resolve();
    };
    handler().catch((err: Error) => {
      console.error(err);
      const message = `Error during initialization: ${err.message}`;
      connection.window.showErrorMessage(message);
    });
  });
});
documents.onDidClose(async change => {
  await connection.sendDiagnostics({ uri: change.document.uri, diagnostics: [] });
});
const linterManager = new LinterManager();
// The content of a text document has changed. This event is emitted
// when the text document first opened or when its content has changed.
async function validateTextDocument(textDocument: TextDocument, fromProjectParser = false) {
  try {
    const vhdlLinter = await linterManager.triggerRefresh(textDocument.uri, textDocument.getText(), projectParser, getDocumentSettings, fromProjectParser);
    const diagnostics = await vhdlLinter.checkAll();
    diagnostics.forEach((diag) => diag.source = 'vhdl-linter');
    await connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
  } catch (err) {
    // Ignore cancelled
    if (!(err instanceof ResponseError && err.code === LSPErrorCodes.RequestCancelled)) {
      throw err;
    }
  }
}

connection.onDidChangeWatchedFiles(() => {
  // Monitored files have change in VS Code
  connection.console.log('We received an file change event');
});
connection.onCodeAction(async (params, token): Promise<CodeAction[]> => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);
  // linter.codeActionEvent.emit()
  const actions = [];
  for (const diagnostic of params.context.diagnostics) {
    if (typeof diagnostic.code === 'number') {
      const callback = linter.diagnosticCodeActionRegistry[diagnostic.code];
      if (typeof callback === 'function') {
        actions.push(...callback(params.textDocument.uri));
      }
    } else if (typeof diagnostic.code === 'string') {
      const codes = diagnostic.code.split(';').map(a => parseInt(a));
      for (const code of codes) {
        const callback = linter.diagnosticCodeActionRegistry[code];
        if (typeof callback === 'function') {
          actions.push(...callback(params.textDocument.uri));
        }

      }
    }
  }
  return actions;
});
connection.onDocumentSymbol(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);

  return DocumentSymbols.get(linter);
});
interface IFindDefinitionParams {
  textDocument: {
    uri: string
  };
  position: Position;
}

const findBestDefinition = async (params: IFindDefinitionParams, token: CancellationToken) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);

  const definitions = findDefinitionLinks(linter, params.position);
  return definitions[0] ?? null;
};

connection.onHover(async (params, token) => {
  const definition = await findBestDefinition(params, token);
  if (definition === null) {
    return null;
  }
  const lines = definition.text.split('\n').slice(definition.targetRange.start.line, definition.targetRange.end.line + 1);
  if (definition.targetRange.start.line === definition.targetRange.end.line) {
    lines[0] = lines[0]!.substring(definition.targetRange.start.character, definition.targetRange.end.character);
  } else {
    lines[0] = lines[0]!.substring(definition.targetRange.start.character);
    lines[lines.length - 1] = lines[lines.length - 1]!.substring(0, definition.targetRange.end.character);
  }
  return {
    contents: {
      language: 'vhdl',
      value: lines.join('\n')
    }
  };
});
connection.onDefinition(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);

  const definitions = findDefinitionLinks(linter, params.position);
  if (definitions.length === 0) {
    return null;
  }
  return definitions;
});
connection.onCompletion(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);
  return getCompletions(linter, params.position);
});
connection.onReferences(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);

  return findReferencesHandler(linter, params.position);

});

connection.onPrepareRename(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);

  return prepareRenameHandler(linter, params.position);
});
connection.onRenameRequest(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);

  return renameHandler(linter, params.position, params.newName);
});
connection.onDocumentFormatting(handleDocumentFormatting);
connection.onFoldingRanges(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);
  return foldingHandler(linter);
});
connection.onDocumentHighlight(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);

  return documentHighlightHandler(linter, params);

});
connection.onWorkspaceSymbol(async params => {
  await initialization;
  const configuration = await getDocumentSettings();
  return workspaceSymbol(params, projectParser, configuration.paths.additional);
});
connection.onSignatureHelp(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);
  return signatureHelp(linter, params.position);
});
connection.languages.semanticTokens.on(async (params, token) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token, false);
  const settings = await getDocumentSettings(new URL(params.textDocument.uri));
  if (!settings.semanticTokens) {
    return {
      data: []
    };
  }
  const tokens = semanticToken(linter, settings.semanticTokensDirectionColoring);
  return tokens;
});
connection.onRequest('vhdl-linter/template', async (params: { textDocument: { uri: string }, type: converterTypes, position?: Position }, token?: CancellationToken) => {
  const linter = await linterManager.getLinter(params.textDocument.uri, token);
  const settings = await getDocumentSettings(new URL(params.textDocument.uri));
  return entityConverter(linter, params.type, settings, params.position);
});
  documents.listen(connection);

  // Listen on the connection
  connection.listen();
