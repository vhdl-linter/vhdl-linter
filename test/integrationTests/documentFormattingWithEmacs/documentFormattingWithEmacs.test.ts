
import { beforeEach, expect, jest, test } from '@jest/globals';
import * as fs from 'fs';
import { pathToFileURL } from 'url';
import { CancellationTokenSource, Position } from 'vscode-languageserver';
import { attachWorkDone } from 'vscode-languageserver/lib/common/progress';
import { documents } from '../../../lib/languageServer';
import { handleDocumentFormatting } from '../../../lib/languageFeatures/documentFormatting';

// eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
const nullProgressReporter = attachWorkDone(undefined as any, /* params */ undefined);
beforeEach(() => {
  jest.clearAllMocks();
});
// jest.mock('vscode-languageserver');
jest.mock('../../../lib/language-server', () => {
  return {
    documents: {
      get: jest.fn()
    },
    connection: {
      window: {
        createWorkDoneProgress: jest.fn(() => ({
          begin: jest.fn(),
          done: jest.fn(),
          report: jest.fn()
        })),
        showErrorMessage: jest.fn()
      }
    }
  };
});
const mockDocuments = jest.mocked(documents);


test('Testing full formatter workflow with emacs', async () => {
  const uri = pathToFileURL(__dirname + '/test_stateMachineCase.vhd');
  const cancellationTokenSource = new CancellationTokenSource();
  mockDocuments.get.mockImplementationOnce(() => {
    return {
      uri: uri.toString(),
      getText: () => fs.readFileSync(uri, {encoding: 'utf8'}),
      languageId: 'vhdl',
      version: 0,
      offsetAt: () => 0,
      positionAt: () => Position.create(0, 0),
      lineCount: 0,

    };
  });
  const formatting = await handleDocumentFormatting({
    textDocument: {
      uri: uri.toString()
    },
    options: {
      tabSize: 2,
      insertSpaces: true
    }
  }, cancellationTokenSource.token, nullProgressReporter);

  expect(formatting).toMatchSnapshot();
});
