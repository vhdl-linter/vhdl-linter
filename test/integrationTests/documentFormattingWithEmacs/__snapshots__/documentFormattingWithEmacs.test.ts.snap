// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing full formatter workflow with emacs 1`] = `
[
  {
    "newText": "library ieee;
use ieee.std_logic_1164.all;
use ieee.numeric_std.all;

entity test_else_generate is
end test_else_generate;

architecture arch of test_else_generate is
  type t_state is (idle, working, workingHard, waiting);
  signal state0 : t_state := idle;
  signal state1 : t_state := idle;
begin

  p_reg : process(all)
  begin
    case state0 is
      when idle =>
        case state1 is
          when idle =>
            state0 <= working;
          when others =>
            state1 <= idle;
        end case;
      when working|workingHard =>
        state0 <= waiting;
      when waiting =>
        state0 <= idle;
      when others =>
        state0 <= workingHard;
    end case;
  end process;

end arch;
",
    "range": {
      "end": {
        "character": 0,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
  },
]
`;
